Sparse vector/matrix: add fast implementation of find_next and find_prev (fixed)

tkluck opened the pull request:
(Opening a new pull request instead of re-opening #23312 because I force-pushed and pull requests don't seem to like that.)

Before this commit, `find_next()` will just use the default implementation of looping over each element. When find_next is called without a function filter as first argument, we *know* that semantics are to find elements x satisfying `x != 0`, so for sparse matrices/vectors, we may only loop over the stored elements.

Some care must be taken for stored zero values; that's the reason for the indirection of `_sparse_find_next` (which only finds the next stored element) and the actual `find_next` (which does actual non-zero checks).

tkluck committed:
Sparse vector/matrix: add fast implementation of find_next and find_prev

Before this commit, find_next() will just use the default implementation
of looping over each element. When find_next is called without a
function filter as first argument, we *know* that semantics are to find
elements x satisfying x != 0, so for sparse matrices/vectors, we may
only loop over the stored elements.

Some care must be taken for stored zero values; that's the reason for
the indirection of _sparse_find_next (which only finds the next stored
element) and the actual find_next (which does actual non-zero checks).

fredrikekre commented:
> (Opening a new pull request instead of re-opening #23312 because I force-pushed and pull requests don't seem to like that.)

For next time, it is fine to leave incomplete PR's open :)

fredrikekre commented on base/sparse/sparsematrix.jl:
```diff
@@ -1332,6 +1332,42 @@ function findnz(S::SparseMatrixCSC{Tv,Ti}) where {Tv,Ti}
+   end
+   row, col = ind2sub(m, i)
+   lo, hi = m.colptr[col], m.colptr[col+1]
+   n = searchsortedfirst(@view(m.rowval[lo:hi-1]), row)
```
You can specify start and stop index to `searchsortedfirst` instead of using a view:
https://github.com/JuliaLang/julia/blob/a945af3aaef528da7813174c17d705b6383f7ea3/base/sort.jl#L160-L172

tkluck replied:
Ooooh good find, thanks. Pushing an update.

tkluck committed:
Sparse findprev/findnext: replace searchsorted(@view...) by searchsorted(...,lo=,hi=)

tkelman commented on base/sparse/sparsematrix.jl:
```diff
@@ -1332,6 +1332,42 @@ function findnz(S::SparseMatrixCSC{Tv,Ti}) where {Tv,Ti}
 end
 
+function _sparse_findnext(m::SparseMatrixCSC, i::Int)
+   if i > length(m)
```
these are slightly under-indented, should be 4 spaces rather than 3

tkelman commented on test/sparse/sparse.jl:
```diff
@@ -1899,3 +1899,26 @@ end
     end
 end
+
+@testset "sparse findprev/findnext operations" begin
```
would be good to include some sparse test arrays with stored zeros

tkluck committed:
_sparse_findnext/prev: fix indentation

(Use git blame -w for finding the non-whitespace edits to
this code.)

tkluck committed:
sparse findprev/findnext: add doctest with stored zeros

tkluck commented:
That AppVeyor failure doesn't seem to be related at all; the code that's failing is
```julia
    tfile = tempname()
    io = open(tfile, "w")
    ccall(:jl_dump_compiles, Void, (Ptr{Void},), io.handle)
    eval(expand(Main, :(for i in 1:10 end)))
    ccall(:jl_dump_compiles, Void, (Ptr{Void},), C_NULL)
    close(io)
    tstats = stat(tfile)
    tempty = tstats.size == 0
    rm(tfile)
    @test tempty == true
```
My guess would be that it's a race condition between `close()` and `stat()`. If so, we may be able to fix it by just calling `stat(io)` instead, which will call `fstat (2)` under the hood.

tkluck commented:
Just opened #23365 for that.

tkluck commented:
Given the resolution of #23365, this should be good to go!

KristofferC commented:
Needs a rebase.

tkluck committed:
Merge branch 'master' into sparse-find-next

tkluck commented:
@KristofferC you got it :) let's see what CI says.

KristofferC commented:
Seems like there is still a conflict here.

tkluck commented:
Oops, merged `master` instead of `origin/master`. Testing+pushing that now.

tkluck committed:
Merge remote-tracking branch 'origin/master' into sparse-find-next

Sacha0 commented:
Linking to #23812 and #23120 in case this pull request need be updated in accord. Best!

tkluck committed:
Merge remote-tracking branch 'origin/master' into sparse-find-next

This fixes a few merge conflicts resulting from other additions to
the sparse codebase.

tkluck committed:
Optimized findnext() for sparse: update now that predicate needs to be explicit

Since we now need explicit predicates [1], this optimization only works
if we know that the predicate is a function that is false for zero
values. As suggested in that pull request, we could find out by calling
`f(zero(eltype(array)))` and hoping that `f` is pure, but I like being
a bit more conservative and only applying this optimization only to the
case where we *know* `f` is equal to `!iszero`.

For clarity, this commit also renames the helper method
_sparse_findnext()  to _sparse_findnextnz(), because now that the
predicate-less version doesn't exist anymore, the `nz` part isn't
implicit anymore either.

[1]: https://github.com/JuliaLang/julia/pull/23812

tkluck commented:
Updated this commit to align with the new semantics for `find` and friends as merged to master from #23812. Very interested to hear your thoughts!

tkluck commented:
CI failures are libgit related and almost surely unrelated to the work in this branch. Should be good to merge!

fredrikekre commented:
@Sacha0 wanna take a look? :)

Sacha0 commented:
> @Sacha0 wanna take a look? :)

I have blocked some time this evening for reviews, and will try to work this pull request in then :). Best!

Sacha0 commented on base/sparse/abstractsparse.jl:
```diff
@@ -28,3 +28,36 @@ function Base.reinterpret(::Type, A::AbstractSparseArray)
+        end
+    end
+    return j
+end
```
Perhaps the following? :)
```julia

function findnext(f::typeof(!iszero), v::AbstractSparseArray, i::Int)
    j = _sparse_findnextnz(v, i)
    while j != 0 && iszero(v[j])
        j = _sparse_findnextnz(v, j+1)
    end
    return j
end
```

tkluck replied:
Looks good! In the interest of expediency, feel free to commit+merge that if you have time.

Sacha0 commented on base/sparse/abstractsparse.jl:
```diff
@@ -28,3 +28,36 @@ function Base.reinterpret(::Type, A::AbstractSparseArray)
+        end
+    end
+    return j
+end
```
Similarly? :)
```julia
function findprev(f::typeof(!iszero), v::AbstractSparseArray, i::Int)
    j = _sparse_findprevnz(v, i)
    while j != 0 && iszero(v[j])
        j = _sparse_findprevnz(v, j-1)
    end
    return j
end
```

Sacha0 commented on base/sparse/sparsevector.jl:
```diff
@@ -737,6 +737,24 @@ function findnz(x::SparseVector{Tv,Ti}) where {Tv,Ti}
+    else
+        return v.nzind[n]
+    end
+end
```
A compact alternative:
```julia
function _sparse_findnextnz(v::SparseVector, i::Int)
    n = searchsortedfirst(v.nzind, i)
    return n <= length(v.nzind) ? v.nzind[n] : 0
end
```

Sacha0 reviewed (approved):
lgtm modulo the minor comments! :)

tkluck committed:
sparse findnext()/findprev(): remove code duplication

Thanks to @Sacha0 for the suggestion!

tkluck commented:
Just updated with your comments @Sacha0 . I hope you don't mind I skipped replacing `if/else/end` by `?:` (which is arguably a matter of taste), but the code duplication removal is an obvious improvement in clarity. Thanks again!

Sacha0 commented on base/sparse/sparsevector.jl:
```diff
@@ -737,6 +737,24 @@ function findnz(x::SparseVector{Tv,Ti}) where {Tv,Ti}
+function _sparse_findnextnz(v::SparseVector, i::Int)
+    n = searchsortedfirst(v.nzind, i)
+    if n > length(v.nzind)
+        return 0
```
For type stability, `return zero(indtype(v))`? (Likewise below.)

Sacha0 reviewed (approved):
Modulo the minor type stability comment above, lgtm! Thanks `@tkluck`! :)

tkluck committed:
sparse findnext()/findprev(): better type stability for exotic index types

This mostly means returning `zero(indtype(...))` instead of `0` in the
not-found case.  In addition, this commit replaces a few `== 0` checks
by `iszero()` to avoid unnecessary promotions. We could similarly
replace `+ 1` by `+ one(...)` but that becomes cumbersome very quickly.

tkluck commented:
Thanks for the review @Sacha0 ! Pushed those updates just now.

tkluck committed:
Merge branch 'master' into sparse-find-next

tkluck commented:
Just added a commit that should resolve the merge conflicts. Do you think we could merge this?

Sacha0 commented:
Looks like! Perhaps @fredrikekre and I should make a brief sweep of the rebased version and then merge?

Sacha0 reviewed (commented):
Test failures appear related?

tkluck commented:
Looks like. Keeping a pull request open for half a year really makes `master` a bit of a moving target 😕 I'm happy to fix it, but what can we do to make sure it gets merged quickly after that?

tkluck committed:
Fix sparse findprev()/findnext() for sub2ind deprecation

This is needed in response to #24715.

fredrikekre commented:
Sorry this has taken so much time, it looks good to me but @Sacha0 should probably take a look and then we should merge!

Sacha0 commented:
If you do not see movement within a few days on a pull request that is approved, passing CI, and not otherwise blocked, please feel welcome to bump! :) Chances are it simply fell off the collective radar and a bump would be appreciated. (Tangentially, having a bot that bumps pull requests under those conditions could work well.)

Sacha0 reviewed (approved):
Thanks for seeing this work through @tkluck! :)
